Post: Extending material theme in Jetpack Compose. (#146)
xmartlabs/blog: _posts/2021-03-07-compose-themes.md

---
layout: post
title: Extending Material theme in Jetpack Compose.
date: 2021-03-06 10:00:00
tags: [Android, Jetpack Compose, Material, Themes, Material Theme]
category: development
author_id: mirland
show: true
featured_image: /images/recap2020/look-back-2020-featured.jpg

> permalink: /blog/extending-material-them-in-jetpack-compose/

Author: ```suggestion
permalink: /blog/extending-material-theme-in-jetpack-compose/
```

---
​
If you are a Jetpack Compose user you may already know that one of its advantages is that it's easier to give your app a consistent look & feel by applying themes.

> Additionally, Material already provides a theme that allows apps based on Material Design to reflect your own product’s branding and styles.

Reviewer: maybe avoid saying material two times in the same sentence
```suggestion
Additionally, Material already provides a theme that allows apps based on it to reflect your own product’s branding and styles.
```

And that is great, but it oftentimes lacks in that it's a bit too strict and limited on what you can and cannot do with it.

> In this post, we'll analyze how to adapt, extend and make more flexible the Material theme so that it can stick to your product's style guidelines.
> ​

Reviewer 2: ```suggestion
In this post, we'll analyze how to adapt, extend and make Metrial Theme more flexible so that it can stick to your product's style guidelines.
​
```

# What's an application theme?
The definition is not complex.

> Themes are a collection of named resources that are useful all throughout a given application.

Reviewer: Maybe omitting the named resources since in the next line you explain that its assigns names to resources?
```suggestion
Themes are a collection of resources that are useful all throughout a given application.
```

More information can be found in [Android's themes documentation], but themes assign semantic names like `colorPrimary` to Android resources, that can be later used as references in different places on the app.
​

> Most Android applications already follow (or should be following) [Material Design Guidelines], and that's why Material created the `MaterialTheme`.

Reviewer 2: ```suggestion
Most Android applications already follow (or should be following) [Material Design Guidelines], and that's why Material created `MaterialTheme`.
```

It's a systematic way to customize Material Design to better reflect your product’s design needs.
Material Theme comprises [color], [typography], and [shape] attributes.
​
​
### Sounds good, but is it perfect?

> Material does a great job, but from my point of view, it's too strict.

Reviewer: a suggestions both ways are ok i think
```suggestion
Material does a great job, but from my point of view, it lacks flexibility.
```

The first one is that your theme will be robust, you can reuse the dimensions in multiple places throughout your app.
The second advantage is that you gain flexibility as you can define custom dimensions based on the device's state or specs.
A practical example could be the dimensions for a small device, where you may have two different sets of dimensions to improve user experience for small devices.
​
```kotlin
@Composable
fun appDims() = if (LocalConfiguration.current.screenWidthDp < 300) {
  smallDeviceAppDims
} else {
  regularAppDims
}
​
​
private val regularAppDims = AppDims(
    textSizeSmall = 12.sp,
    textSizeMedium = 14.sp,
    textSizeRegular = 16.sp,
​
    listItemVerticalPadding = 27.dp,
    listItemHorizontalPadding = 30.dp,
)
​
private val smallDeviceAppDims = AppDims(
    textSizeSmall = 12.sp,
    textSizeMedium = 13.sp,
    textSizeRegular = 14.sp,
​
    listItemVerticalPadding = 15.dp,
    listItemHorizontalPadding = 14.dp,
)
```
​
Then, you have to add the `AppDims` to your `AppTheme` just like we did with the `AppColors` example.
You can check the [final code here].
​
​
# Conclusions
UI/UX is one of the most important things in a mobile app.

> Material has some guides to help improve that and it also allows to have consistency wit other apps on the same platform.

Reviewer 2: ```suggestion
Material has some guides to help improve that and it also allows to have consistency with other apps on the same platform.
```

> Themes allow for consistency throughout our own application, especially if you define it at the beginning, just by applying minor configurations.

Reviewer 2: ```suggestion
Themes allow for consistency throughout an application, especially if you define it at the beginning, just by applying minor configurations.
```

> However, we saw that Material theme is not really all that flexible, and if designers don't stick 100% to the their guidelines, you may be in trouble as you may now have a mix of Material and other theme not really related to it.

Reviewer: ```suggestion
However, we saw that Material theme lacks flexibility, and if designers don't stick 100% to the their guidelines, you may be in trouble as you may now have a mix of Material and other theme not really related to it.
```

> However, we saw that the Material theme is not really all that flexible, and if designers don't stick 100% to the their guidelines, you may be in trouble as you may now have a mix of Material and other theme not really related to it.

Reviewer 2: ```suggestion
However, we saw that Material theme is not really all that flexible, and if designers don't stick 100% to the their guidelines, you may be in trouble as you may now have a mix of Material and other theme not really related to it.
```

In this post we presented some ideas to avoid these issues, a way to extend the Material theme while making it flexible, and the most important thing, adaptable to your product.

> [The code in this blog should be useful for most applications, but I recommend you check out the full implementation on GitHub], and adapt the general ideas to your specific use case.

Reviewer: maybe change this link to just github?
```suggestion
The code in this blog should be useful for most applications, but I recommend you check out the full implementation on [GitHub], and adapt the general ideas to your specific use case.
```

> [The code in this blog should be useful for most application, I recommend you check the full implementation on GitHub], and adapt the general ideas to your specific use case.

Reviewer 2: ```suggestion
[The code in this blog should be useful for most applications, but I recommend you check out the full implementation on GitHub], and adapt the general ideas to your specific use case.
```

---
layout: post

> title: Extending Material theme in Jetpack Compose.

Reviewer 3: ```suggestion
title: Extending Material theme in Jetpack Compose
```

date: 2021-03-06 10:00:00
tags: [Android, Jetpack Compose, Material, Themes, Material Theme]
category: development
author_id: mirland
show: true
featured_image: /images/recap2020/look-back-2020-featured.jpg
permalink: /blog/extending-material-theme-in-jetpack-compose/
---
​

> If you are a Jetpack Compose user you may already know that one of its advantages is that it's easier to give your app a consistent look & feel by applying themes.

Reviewer 3: Also for context, what about adding some reference link here to Jetpack Compose?

Author: What about something like:

```suggestion
If you have tried [Jetpack Compose], the new Android's modern toolkit for building UIs, you may already know that one of its advantages is that it's easier to give your app a consistent look & feel by applying themes.
```

?

In a few words, it add context and some documentation

Reviewer 3: I like it!

Reviewer 3: What about "building UIs" -> "building mobile user interfaces"? As it's the very first sentence.

Reviewer 3: Oh, and having both "new" and "modern" is redundant.

Author: > building mobile user interfaces"

It's the new Android's toolkit, but it's not just jor android, it's unbundled, it's available on desktop and web (both are on alpha).
So, maybe just:

```suggestion
If you have tried [Jetpack Compose], the new Android's toolkit for building UIs, you may already know that one of its advantages is that it's easier to give your app a consistent look & feel by applying themes.
```
My point here, it applies to web and desktop, but anybody should be using it

Reviewer 3: maybe: "the new framework for building user interfaces for Android ([and beta for Web and Desktop]), you may ...", with a link ?

Reviewer 3: Also: "you may already know that one of its advantages is that it's easier to give your app a consistent look & feel by applying themes." -> "you may know it's now easier to ..."

And I personally don't like the ampersand there, I'd use and. But I'm not 100% sure.

> If you have tried [Jetpack Compose], the new framework for building user interfaces for Android (and in early steps for [Desktop](https://www.jetbrains.com/lp/compose/) and [Web](https://blog.jetbrains.com/kotlin/2021/05/technology-preview-jetpack-compose-for-web/)), you may know it's now easier to give your app a consistent look and feel by applying themes.

Reviewer 4: ```suggestion
Building Android UIs with a consistent look and feel using the new [Jetpack Compose] framework it's easier than ever by setting up themes.
```

> Additionally, Material already provides a theme that allows apps based on it to reflect your own product's branding and styles.
> If you haven't tried it, I'd recommend you to check out [Android's documentation](https://developer.android.com/jetpack/compose/themes), it's very simple and clear how you can use it in your app.
> Material theme is great, but it oftentimes lacks in that it's a bit too strict and limited on what you can and cannot do with it.

Reviewer 4: ```suggestion
Compose offers an implementation of the Material design that makes it easy to use Material style throughout the whole app just by adding few lines of code. As you may already know, Material theme allows us to have a consistent and uniform UI that follows Android UI best practices and guidelines, but at the same time, it's super rigid and limited when the app needs a more complex UI customization level.
```

Author: I don't like the idea of losing the info that here we will not explain how to use the material theme, it's something that you should read before this blog. I think it's important

Author: ```suggestion
Compose offers an implementation of the Material design that makes it easy to use Material style throughout the whole app just by adding few lines of code.
If you haven't tried it, you can check out [Android's documentation](https://developer.android.com/jetpack/compose/themes).
As you may already know, Material theme allows us to have a consistent and uniform UI that follows Android UI best practices and guidelines, but at the same time, it's super rigid and limited when the app needs a more complex UI customization level.
```

​
# What's an application theme?
The definition is not complex.
Themes are a collection of resources that are useful throughout a given application.

> More information can be found in [Android's themes documentation], but themes assign semantic names like `colorPrimary` to Android resources, that can be later used as references in different places on the app.

Reviewer 3: ```suggestion
More information can be found in [Android's themes documentation], but themes assign semantic names such as `colorPrimary` to Android resources, that can be later used as references in different places on the app.
```

"such as" should go for real examples. "like" is more like for figurative or related things.

> In this post, we'll not talk about how can you apply the Material theme in your app, we'll analyze how to adapt, customize, extend and make Material Theme more flexible so that it can stick to your product's style guidelines.

Reviewer 4: ```suggestion
In this post, we'll learn how to adapt, customize and extend Compose Material Theme in order to fulfill your app UI requirement and get rid of material limitations. 
```

> The definition is not complex.

Reviewer 4: ```suggestion
```

> Themes are a collection of resources that are useful throughout a given application.
> More information can be found in [Android's themes documentation], but themes assign semantic names such as `colorPrimary` to Android resources, that can be later used as references in different places on the app.

Reviewer 4: ```suggestion
[Android Theme] is just a collection of resources that are useful throughout a given application. Each resource has a semantic name such as `colorPrimary` that can be used as references from different places on the app.
```

> Most Android applications already follow (or should be following) [Material Design Guidelines], and that's why Material created `MaterialTheme`.

Reviewer 4: ```suggestion
Most Android apps already follow [Material Design Guidelines], and that's the reason Android SDK already provides the `MaterialTheme` for Compose apps.
```

Author: it's not the android sdk... it super confusing 
```suggestion
Most Android apps already follow [Material Design Guidelines], and that's the reason Jetpack Compose already provides the `MaterialTheme` for Compose apps.
```

It's a systematic way to customize Material Design to better reflect your product’s design needs.
Material Theme comprises [color], [typography], and [shape] attributes.

### Sounds good, but is it perfect?

> Material does a great job, but from my point of view, it lacks flexibility.

Reviewer 4: ```suggestion
Material does a great job making sure every app follows android UI guidelines, but it clearly lacks flexibility.
```

> So, here lies the **first problem**: you have to adapt your look and feel to the Material guidelines, and if you don't, then it becomes hard to use it.

Reviewer 4: So, here lies the **first problem**: Material theme by itself it's not flexible enough to build complex app look and feels.

Author: if we don't use bold in the next sentence, we should remove it here too
```suggestion
So, here lies the first problem: you have to adapt your look and feel to the Material guidelines, and if you don't, then it becomes hard to use it.
```

> The **second issue** is that Material comprises only [color], [typography], and [shape] attributes, but what if we want to define more resources in a theme, such as dimensions or icons?

Reviewer 4: ```suggestion
Secondly, Material comprises only [color], [typography], and [shape] attributes, but what if your app needs more resources, such as dimensions or icons?
```

# Extending Material Colors

> Compose provides the [`Colors`] class to model the [Material color system].
> It's a class that defines the 9 main Material colors, and it also provides two functions to define both the light and dark color sets.

Reviewer 4: ```suggestion
Compose provides the [`Colors`] class to model the [Material color system] which includes nine colors and two functions to define light and dark color sets.
```

> In my experience, 9 colors are not enough, as design teams usually define many more.

Reviewer 4: ```suggestion
According to our experience, most apps need way more than 9 colors.
```

> In my experience, 9 colors are not enough, as design teams usually define many more.

Reviewer 4: El `as design teams usually define many more.` es un detalle que se puede omitir, en mi opinipn no aporta nada.

  val colors: AppColors
    @Composable
    @ReadOnlyComposable
    get() = LocalAppColors.current
}
​
private val LocalAppColors = staticCompositionLocalOf {
  defaultAppColors() // For instance, pink dark colors
}
​
@Composable
fun AppTheme(
    darkTheme: Boolean = isSystemInDarkTheme(),
    colorPalette: AppColorPalette = AppColorPalette.PINK,
    content: @Composable () -> Unit,
) {
  val colors = appColors(colorPalette = colorPalette, darkTheme = darkTheme)
  CompositionLocalProvider(
      LocalAppColors provides colors,
  ) {
    MaterialTheme(
        colors = colors.materialColors,
        content = content,
    )
  }
}
```
​
These lines define two things, a Composable app theme to handle our custom colors and a static class that provides them.
The full implementation of these classes can be found [here](https://github.com/xmartlabs/gong/blob/b0b617e56403c1f499704111acad89093aa3c9d6/app/src/main/java/com/xmartlabs/gong/ui/theme/AppTheme.kt#L10).
​
​
### Extending [shape] and [typography].
​
To extend these classes, we can follow the same idea, we can create a [`AppShapes`] class and a [`AppTypographies`] class the same way and declare custom theme properties.
​

> I will not explain the code because it's analog, you can find it on [GitHub](https://github.com/xmartlabs/gong/tree/b0b617e56403c1f499704111acad89093aa3c9d6/app/src/main/java/com/xmartlabs/gong/ui/theme) and if you have any questions, don't forget to post a comment!

Author: I can say it's similar to the previous one. 
In this case, what's better, similar, or consistent?

> I will not explain the code because it's analog, you can find it on [GitHub](https://github.com/xmartlabs/gong/tree/b0b617e56403c1f499704111acad89093aa3c9d6/app/src/main/java/com/xmartlabs/gong/ui/theme) and if you have any questions, don't forget to post a comment!

Reviewer 3: "similar to the previous one"?

# Adding custom resources to your theme

> In the previous section, we learned how to extend the properties provided by Material, but what if we wanted to define new ones?

Reviewer 4: ```suggestion
So far, we've learned how to extend the properties provided by Material, but what if we wanted to define new ones?
```

> Remember, a theme is a semantic resource set, so we could also have dimensions, icons, or any other resource that will be used in the app.

Reviewer 4: ```suggestion
Remember, a theme is a semantic resource set, so we could also need dimensions, icons, or any other resource to build the custom app UI.
```

## Defining dimensions in our theme

> I usually define dimensions in my theme since many of them have a meaning for me.

Reviewer 4: No me gusta el uso de `I`, por momentos es poco inclusivo con el equipo.

Author: maybe: 
```suggestion
Dimensions are something that we usually define in our theme since many of them have a meaning in our app context.
```

​
```kotlin
@Composable
fun appDims() = if (LocalConfiguration.current.screenWidthDp < 300) {
  smallDeviceAppDims
} else {
  regularAppDims
}
​
​
private val regularAppDims = AppDims(
    textSizeSmall = 12.sp,
    textSizeMedium = 14.sp,
    textSizeRegular = 16.sp,
​
    listItemVerticalPadding = 27.dp,
    listItemHorizontalPadding = 30.dp,
)
​
private val smallDeviceAppDims = AppDims(
    textSizeSmall = 12.sp,
    textSizeMedium = 13.sp,
    textSizeRegular = 14.sp,
​
    listItemVerticalPadding = 15.dp,
    listItemHorizontalPadding = 14.dp,
)
```
​
Then, you have to add the `AppDims` to your `AppTheme` just like we did with the `AppColors` example.
You can check the [final code here](https://github.com/xmartlabs/gong/blob/b0b617e56403c1f499704111acad89093aa3c9d6/app/src/main/java/com/xmartlabs/gong/ui/theme/AppDims.kt).
​
​
# Conclusions
UI/UX is one of the most important things in a mobile app.
Material has some guides to help improve that and it also allows to have consistency with other apps on the same platform.
Themes allow for consistency throughout an application, especially if you define it at the beginning, just by applying minor configurations.

> However, we saw that Material theme lacks flexibility, and if designers don't stick 100% to their guidelines, you may be in trouble as you may now have a mix of Material and another theme not related to it.

Author: What this phrase try to tell is: if you use the material theme without these recommendations, you may have the material theme and a lot of separate theme components (that are not handled by your app theme) in your app. You have to mix both and it will be hard to maintain (and the code may be horrible).

> you may be in trouble as you will have to maintain multiple themes.

Your app theme will be divided into multiple themes (or worst, only the material theme and a lot of separate components in your app). However, is it clear?

> However, we saw that Material theme lacks flexibility, and if designers don't stick 100% to their guidelines, you may be in trouble as you may now have a mix of Material and another theme not related to it.

Reviewer 3: Yeah. Maybe say "as you will have to maintain many separate theme components not handled directly by your app theme"?

Each application can also define multiple color palettes, often selecting one or another depending on the app's state.
Suppose that we now want to define two color palettes, one based on blue colors and the other based on pink ones:

```kotlin
enum class AppColorPalette {
  BLUE,
  PINK,
}

// AppCustomColors is private because we shouldn't expose a color that is not in a theme
private object AppCustomColors {
  val PINK_AMARANTH = Color(0xffe91e63)
  val PINK_MAROON = Color(0xffc2185b)
  val PINK_MAUVELOUS = Color(0xfff48fb0)
}

private val darkPinkColors = AppColors(
    linkTextColor = AppCustomColors.PINK_AMARANTH,
    materialColors = darkColors(
        primary = AppCustomColors.PINK_MAUVELOUS,
        primaryVariant = AppCustomColors.PINK_MAROON,
        .... // Material Colors
    )

// Define lightPinkColors, darkBlueColors, lightBlueColors the same way

@Composable
fun appColors(colorPalette: AppColorPalette, darkTheme: Boolean): AppColors =
    when (colorPalette) {
      AppColorPalette.BLUE -> if (darkTheme) darkBlueColors else lightBlueColors
      AppColorPalette.PINK -> if (darkTheme) darkPinkColors else lightPinkColors
    }

> 
> ```

Reviewer 3: ```suggestion
```
```

So the rendered code block just fits the code :smile:

To extend these classes, we can follow the same idea.
We create an [`AppShapes`] class and an [`AppTypographies`] class the same way and declare custom theme properties.

I will not explain the code because it's , you can also find it on [Gong repo](https://github.com/xmartlabs/gong/tree/b0b617e/app/src/main/java/com/xmartlabs/gong/ui/theme) and if you have any questions, don't forget to comment!

# Adding custom resources to your theme

In the previous section, we learned how to extend the properties provided by Material, but what if we wanted to define new ones?
Remember, a theme is a semantic resource set, so we could also have dimensions, icons, or any other resource that will be used in the app.

## Defining dimensions in our theme

I usually define dimensions in my theme since many of them have a meaning for me.
Examples are list items' padding, a "small size", and a container margin.

We can use a similar approach here and define a dimension class:

```kotlin
@Immutable
data class AppDims(
    val textSizeSmall: TextUnit,
    val textSizeMedium: TextUnit,
    val textSizeRegular: TextUnit,

    val listItemVerticalPadding: Dp,
    val listItemHorizontalPadding: Dp,
)

> ```

Reviewer 3: Remove this new line so the rendered code block just fits the text :sweat_smile:
